sdk/python/kubeflow/training/utils/utils.py

```diff
@@ -330,32 +326,45 @@ def get_pytorchjob_template(
         ),
     )
 
-    if num_procs_per_worker > 0:
+    if num_procs_per_worker:
         pytorchjob.spec.nproc_per_node = str(num_procs_per_worker)
     if elastic_policy:
         pytorchjob.spec.elastic_policy = elastic_policy
 
+    # Create Master replica if that is set.
     if master_pod_template_spec:
         pytorchjob.spec.pytorch_replica_specs[
             constants.REPLICA_TYPE_MASTER
         ] = models.KubeflowOrgV1ReplicaSpec(
             replicas=1,
             template=master_pod_template_spec,
         )
+    # If we don't define Master template, use the Worker template.
+    else:
```

Comment:
This is bit confusing to me.  Typically, master is explictly set when it is required by the user. User sets num_workers but we use master replica with workers=num_workers-1 This fix is bit contounter intuituve.

Reply:
If master is mandatory, we can add `needsMaster` flag in train() which is defaulted to True.  And add a validation check to prevent user overriding it. (for now)

Reply:
>Typically, master is explictly set when it is required by the user.

@johnugeorge As I said, currently our controller doesn't set appropriate ENV variables if Master is not set in PyTorchJob. Without RANK and WORLD_SIZE we just can't run PyTorch DDP properly.

>User sets num_workers but we use master replica with workers=num_workers-1 This fix is bit contounter intuituve.

That is because Master is the Worker with `RANK=0`. It just work like a coordinator across all workers.

>If master is mandatory, we can add needsMaster flag in train() which is defaulted to True.

Why do we need to add this flag to the `train()` func if that is not necessary for user ? User just need to define how many workers they want to use and we are going to create PyTorchJob with appropriate spec.

In the future, if we are going to support creation of PyTorchJob without Master replica, we can modify the SDK.

Reply:
@andreyvelich 
I am able to run the following spec with only workers when using elastic policy.
```
apiVersion: "kubeflow.org/v1"
kind: "PyTorchJob"
metadata:
  name: "torchrun-test"
  namespace: training
spec:
  nprocPerNode: "1"
  elasticPolicy:
    rdzvBackend: c10d
    # nProcPerNode: 2
  pytorchReplicaSpecs:
    worker:
      replicas: 2
      restartPolicy: OnFailure
      template:
        metadata:
          labels:
            kind: "deepanker2"
          annotations:
            test_vesrion: "1"
            sidecar.istio.io/inject: "false"
        spec:
          containers:
            - name: pytorch
              image: quay.io/deepanker_gupta/kubeflow_training:torchrun
              imagePullPolicy: Always
              args: ["torchrun","/workspace/exp/training.py"]
              resources: 
                limits:
                  nvidia.com/gpu: 1
                  cpu: 10
                  memory: '10Gi'
```

You can try it running on your cluster as the image is public.

Reply:
the following environment variables were getting set when I ran the above

```
environ({'PATH': '/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin', 'HOSTNAME': 'torchrun-test-worker-1', 'NVIDIA_VISIBLE_DEVICES': 'GPU-28067a3b-25cd-58ed-216b-d315c6c9aeff', 'NVIDIA_DRIVER_CAPABILITIES': 'compute,utility', 'LD_LIBRARY_PATH': '/usr/local/nvidia/lib:/usr/local/nvidia/lib64', 'PYTORCH_VERSION': '2.0.1', 'PYTHONUNBUFFERED': '1', 'PET_RDZV_ENDPOINT': 'torchrun-test-worker-0:23456', 'PET_RDZV_BACKEND': 'c10d', 'PET_NNODES': '2:2', 'KUBERNETES_PORT': 'tcp://172.19.0.1:443', 'KUBERNETES_PORT_443_TCP': 'tcp://172.19.0.1:443', 'KUBERNETES_PORT_443_TCP_PROTO': 'tcp', 'KUBERNETES_PORT_443_TCP_PORT': '443', 'KUBERNETES_PORT_443_TCP_ADDR': '172.19.0.1', 'KUBERNETES_SERVICE_HOST': '172.19.0.1', 'KUBERNETES_SERVICE_PORT': '443', 'KUBERNETES_SERVICE_PORT_HTTPS': '443', 'HOME': '/root', 'LC_CTYPE': 'C.UTF-8', 'LOCAL_RANK': '0', 'RANK': '1', 'GROUP_RANK': '1', 'ROLE_RANK': '1', 'ROLE_NAME': 'default', 'LOCAL_WORLD_SIZE': '1', 'WORLD_SIZE': '2', 'GROUP_WORLD_SIZE': '2', 'ROLE_WORLD_SIZE': '2', 'MASTER_ADDR': 'torchrun-test-worker-0', 'MASTER_PORT': '45614', 'TORCHELASTIC_RESTART_COUNT': '0', 'TORCHELASTIC_MAX_RESTARTS': '0', 'TORCHELASTIC_RUN_ID': 'none', 'TORCHELASTIC_USE_AGENT_STORE': 'False', 'NCCL_ASYNC_ERROR_HANDLING': '1', 'TORCHELASTIC_ERROR_FILE': '/tmp/torchelastic_i2rkd3p1/none_1p31jibu/attempt_0/0/error.json'})
```

Reply:
@deepanker13 That's right. That is because when you run PyTorch with `torchrun` it automatically sets RANK and WORLD_SIZE env: https://pytorch.org/docs/stable/elastic/run.html#module-torch.distributed.run.
But if user just uses the `nn.parallel.DistributedDataParallel` package to set the PyTorch DDP (as in [this example](https://github.com/kubeflow/training-operator/blob/master/examples/sdk/create-pytorchjob-from-func.ipynb)), we have to setup RANK and WORLD_SIZE for each pod.

In the future, we can configure our PyTorchJob to only run with `torchrun`, but it requires changes in SDK and Controller.

Reply:
@andreyvelich you are correct that the job fails when using only workers when we are using the spec without elastic policy. For example the below spec is failing:-

```
apiVersion: "kubeflow.org/v1"
kind: "PyTorchJob"
metadata:
  name: "torchrun-test"
  namespace: training
spec:
  nprocPerNode: "1"
  pytorchReplicaSpecs:
    Worker:
      replicas: 2
      restartPolicy: OnFailure
      template:
        metadata:
          labels:
            kind: "deepanker2"
          annotations:
            test_vesrion: "1"
            sidecar.istio.io/inject: "false"
        spec:
          containers:
            - name: pytorch
              image: quay.io/deepanker_gupta/kubeflow_training:torchrun
              imagePullPolicy: Always
              args: ["torchrun","/workspace/exp/training.py"]
              resources: 
                limits:
                  nvidia.com/gpu: 1
                  cpu: 10
                  memory: '10Gi'
```